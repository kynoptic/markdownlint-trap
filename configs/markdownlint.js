// @ts-check

"use strict";

/**
 * Programmatic markdownlint configuration with custom rules
 *
 * @description This module provides a ready-to-use markdownlint configuration object
 * that can be imported directly in JavaScript-based markdownlint configurations.
 * It enables the custom rules defined in this package while disabling some default rules.
 *
 * @module markdownlint-config
 * @example
 * // In your JavaScript file
 * const markdownlintConfig = require("markdownlint-clarity/markdownlint");
 *
 * // Use the configuration
 * const markdownlint = require("markdownlint");
 * const result = markdownlint.sync({
 *   files: ["README.md"],
 *   config: markdownlintConfig
 * });
 */

const sentenceCase = require("./rules/sentence-case.js");
const backtickCodeElements = require("./rules/backtick-code-elements.js");

module.exports = {
  default: true,
  MD204: false, // Code blocks should use fenced syntax
  MD013: false, // Line length
  MD024: false, // Multiple headings with the same content
  customRules: [sentenceCase, backtickCodeElements],
  "sentence-case-headings-bold": {
    enabled: true,
  },
  "backtick-code-elements": {
    enabled: true,
  },
};
