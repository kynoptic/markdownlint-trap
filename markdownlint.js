"use strict";

const sentenceCase = require("./rules/sentence-case.js");
const backtickCodeElements = require("./rules/backtick-code-elements.js");

module.exports = {
  "default": true,
  "MD204": false,
  "MD013": false,
  "customRules": [
    sentenceCase,
    backtickCodeElements
  ],
  "sentence-case-headings-bold": {
    "enabled": true
  },
  "backtick-code-elements": {
    "enabled": true
  }
};
