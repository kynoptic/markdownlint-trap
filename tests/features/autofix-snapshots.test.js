/**
 * @integration
 * Snapshot tests for autofix functionality across all rules.
 * These tests capture the exact transformations made by autofix logic.
 */
import path from "path";
import fs from "fs";
import { fileURLToPath } from "url";
import { describe, test, expect } from "@jest/globals";
import { lint } from "markdownlint/promise";
import { applyFixes } from "markdownlint";
import sentenceRule from "../../src/rules/sentence-case-heading.js";
import backtickRule from "../../src/rules/backtick-code-elements.js";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

/**
 * Apply autofix transformations to content and return the result.
 * @param {string} content - Original markdown content
 * @param {object} rule - The markdownlint rule to apply
 * @returns {Promise<string>} - The fixed content
 */
async function applyAutofixToContent(content, rule) {
  const results = await lint({
    customRules: [rule],
    strings: {
      content: content
    },
    resultVersion: 3,
    fix: true,
  });

  const violations = results.content || [];
  const fixes = violations.filter(v => v.fixInfo);
  
  if (fixes.length === 0) {
    return content; // No fixes applied
  }

  return applyFixes(content, fixes);
}

/**
 * Create a snapshot test for autofix functionality.
 * @param {string} testName - Name of the test
 * @param {string} fixturePath - Path to the fixture file
 * @param {object} rule - The markdownlint rule to test
 */
function createAutofixSnapshotTest(testName, fixturePath, rule) {
  test(`${testName} autofix snapshot`, async () => {
    const originalContent = fs.readFileSync(fixturePath, "utf8");
    const fixedContent = await applyAutofixToContent(originalContent, rule);
    
    // Create a snapshot object that shows both before and after
    const snapshot = {
      original: originalContent,
      fixed: fixedContent,
      hasChanges: originalContent !== fixedContent
    };
    
    expect(snapshot).toMatchSnapshot();
  });
}

describe("Autofix snapshot tests", () => {
  describe("sentence-case-heading rule", () => {
    const sentenceCaseFixturePath = path.join(
      __dirname,
      "../fixtures/sentence-case/autofix.fixture.md"
    );
    
    createAutofixSnapshotTest(
      "sentence-case-heading",
      sentenceCaseFixturePath,
      sentenceRule
    );
  });

  describe("backtick-code-elements rule", () => {
    const backtickFixturePath = path.join(
      __dirname,
      "../fixtures/backtick/autofix.fixture.md"
    );
    
    createAutofixSnapshotTest(
      "backtick-code-elements",
      backtickFixturePath,
      backtickRule
    );

    const backtickMinimalFixturePath = path.join(
      __dirname,
      "../fixtures/backtick/autofix-minimal.fixture.md"
    );
    
    createAutofixSnapshotTest(
      "backtick-code-elements-minimal",
      backtickMinimalFixturePath,
      backtickRule
    );
  });

  describe("Individual autofix transformations", () => {
    test("sentence case heading transformations", async () => {
      const testCases = [
        "# This Is A Title Case Heading",
        "## API Documentation For Users", 
        "### Working With Node.js And Express",
        "# 1. Getting Started With Development",
        "## 2. API Rate Limiting Best Practices",
        "# CSS and HTML best practices",
        "## Using JSON with REST APIs",
        "### The HTTP and HTTPS protocols"
      ];

      for (const testCase of testCases) {
        const fixed = await applyAutofixToContent(testCase, sentenceRule);
        expect({
          original: testCase,
          fixed: fixed,
          hasChanges: testCase !== fixed
        }).toMatchSnapshot();
      }
    });

    test("backtick code elements transformations", async () => {
      const testCases = [
        "Use the package.json file to configure dependencies.",
        "Run npm install to install packages.",
        "The src/index.js file contains the main logic.",
        "Set the NODE_ENV variable to production.",
        "Check the README.md for installation instructions.",
        "Navigate to the /usr/local/bin directory.",
        "Execute the git status command to check changes.",
        "The .gitignore file excludes temporary files."
      ];

      for (const testCase of testCases) {
        const fixed = await applyAutofixToContent(testCase, backtickRule);
        expect({
          original: testCase,
          fixed: fixed,
          hasChanges: testCase !== fixed
        }).toMatchSnapshot();
      }
    });

    test("edge cases with no changes", async () => {
      const noChangeCases = [
        "# This is already correct", // Sentence case heading
        "Use the `package.json` file correctly.", // Already has backticks
        "## API documentation", // Proper acronym usage
        "### Working with Node.js properly" // Proper technical term
      ];

      for (const testCase of noChangeCases) {
        const sentenceFixed = await applyAutofixToContent(testCase, sentenceRule);
        const backtickFixed = await applyAutofixToContent(testCase, backtickRule);
        
        expect({
          original: testCase,
          sentenceFixed: sentenceFixed,
          backtickFixed: backtickFixed,
          sentenceHasChanges: testCase !== sentenceFixed,
          backtickHasChanges: testCase !== backtickFixed
        }).toMatchSnapshot();
      }
    });
  });
});