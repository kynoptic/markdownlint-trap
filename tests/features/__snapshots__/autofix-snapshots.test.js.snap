// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autofix snapshot tests Individual autofix transformations backtick code elements transformations 1`] = `
{
  "fixed": "Use the \`package.json\` file to configure dependencies.
",
  "hasChanges": true,
  "original": "Use the package.json file to configure dependencies.",
}
`;

exports[`Autofix snapshot tests Individual autofix transformations backtick code elements transformations 2`] = `
{
  "fixed": "Run \`npm install\` to install packages.
",
  "hasChanges": true,
  "original": "Run npm install to install packages.",
}
`;

exports[`Autofix snapshot tests Individual autofix transformations backtick code elements transformations 3`] = `
{
  "fixed": "The src/\`index.js\` file contains the main logic.
",
  "hasChanges": true,
  "original": "The src/index.js file contains the main logic.",
}
`;

exports[`Autofix snapshot tests Individual autofix transformations backtick code elements transformations 4`] = `
{
  "fixed": "Set the \`NODE_ENV\` variable to production.
",
  "hasChanges": true,
  "original": "Set the NODE_ENV variable to production.",
}
`;

exports[`Autofix snapshot tests Individual autofix transformations backtick code elements transformations 5`] = `
{
  "fixed": "Check the \`README.md\` for installation instructions.
",
  "hasChanges": true,
  "original": "Check the README.md for installation instructions.",
}
`;

exports[`Autofix snapshot tests Individual autofix transformations backtick code elements transformations 6`] = `
{
  "fixed": "Navigate to the /\`usr/local/bin\` directory.
",
  "hasChanges": true,
  "original": "Navigate to the /usr/local/bin directory.",
}
`;

exports[`Autofix snapshot tests Individual autofix transformations backtick code elements transformations 7`] = `
{
  "fixed": "Execute the \`git status\` command to check changes.
",
  "hasChanges": true,
  "original": "Execute the git status command to check changes.",
}
`;

exports[`Autofix snapshot tests Individual autofix transformations backtick code elements transformations 8`] = `
{
  "fixed": "The \`.gitignore\` file excludes temporary files.
",
  "hasChanges": true,
  "original": "The .gitignore file excludes temporary files.",
}
`;

exports[`Autofix snapshot tests Individual autofix transformations edge cases with no changes 1`] = `
{
  "backtickFixed": "# This is already correct
",
  "backtickHasChanges": true,
  "original": "# This is already correct",
  "sentenceFixed": "# This is already correct
",
  "sentenceHasChanges": true,
}
`;

exports[`Autofix snapshot tests Individual autofix transformations edge cases with no changes 2`] = `
{
  "backtickFixed": "Use the \`package.json\` file correctly.
",
  "backtickHasChanges": true,
  "original": "Use the \`package.json\` file correctly.",
  "sentenceFixed": "Use the \`package.json\` file correctly.
",
  "sentenceHasChanges": true,
}
`;

exports[`Autofix snapshot tests Individual autofix transformations edge cases with no changes 3`] = `
{
  "backtickFixed": "## API documentation
",
  "backtickHasChanges": true,
  "original": "## API documentation",
  "sentenceFixed": "## API documentation
",
  "sentenceHasChanges": true,
}
`;

exports[`Autofix snapshot tests Individual autofix transformations edge cases with no changes 4`] = `
{
  "backtickFixed": "### Working with Node.js properly
",
  "backtickHasChanges": true,
  "original": "### Working with Node.js properly",
  "sentenceFixed": "### Working with Node.js properly
",
  "sentenceHasChanges": true,
}
`;

exports[`Autofix snapshot tests Individual autofix transformations sentence case heading transformations 1`] = `
{
  "fixed": "# This is a title case heading
",
  "hasChanges": true,
  "original": "# This Is A Title Case Heading",
}
`;

exports[`Autofix snapshot tests Individual autofix transformations sentence case heading transformations 2`] = `
{
  "fixed": "## API Documentation for users
",
  "hasChanges": true,
  "original": "## API Documentation For Users",
}
`;

exports[`Autofix snapshot tests Individual autofix transformations sentence case heading transformations 3`] = `
{
  "fixed": "### Working with Node.js and express
",
  "hasChanges": true,
  "original": "### Working With Node.js And Express",
}
`;

exports[`Autofix snapshot tests Individual autofix transformations sentence case heading transformations 4`] = `
{
  "fixed": "# 1. getting started with development
",
  "hasChanges": true,
  "original": "# 1. Getting Started With Development",
}
`;

exports[`Autofix snapshot tests Individual autofix transformations sentence case heading transformations 5`] = `
{
  "fixed": "## 2. API rate limiting best practices
",
  "hasChanges": true,
  "original": "## 2. API Rate Limiting Best Practices",
}
`;

exports[`Autofix snapshot tests Individual autofix transformations sentence case heading transformations 6`] = `
{
  "fixed": "# CSS and HTML best practices
",
  "hasChanges": true,
  "original": "# CSS and HTML best practices",
}
`;

exports[`Autofix snapshot tests Individual autofix transformations sentence case heading transformations 7`] = `
{
  "fixed": "## Using JSON with REST APIs
",
  "hasChanges": true,
  "original": "## Using JSON with REST APIs",
}
`;

exports[`Autofix snapshot tests Individual autofix transformations sentence case heading transformations 8`] = `
{
  "fixed": "### The HTTP and HTTPS protocols
",
  "hasChanges": true,
  "original": "### The HTTP and HTTPS protocols",
}
`;

exports[`Autofix snapshot tests backtick-code-elements rule backtick-code-elements autofix snapshot 1`] = `
{
  "fixed": "# Auto-fix fixture for backtick-code-elements

## Multiple code elements on the same line

Run \`npm install\` to set up your project and then install debug@latest. <!-- ❌ -->

## File paths and commands

Look in the \`src/utils\` directory and run \`git clone\` <https://github.com/user/repo.git> <!-- ❌ -->

## Multiple elements spread across the document

Edit \`config.yaml\` in the settings/ folder <!-- ❌ -->

Run ./\`build.sh\` to compile the project <!-- ❌ -->

Add \`import os\` at the top of your script <!-- ❌ -->

## Nested code elements (should only fix outer ones)

The command \`echo PATH=/usr/bin\` sets your path variable to /\`usr/bin\` <!-- ❌ -->

## Elements after punctuation

Open \`readme.md\`, \`config.ini\`, and \`.env\` for configuration. <!-- ❌ -->
",
  "hasChanges": true,
  "original": "# Auto-fix fixture for backtick-code-elements

## Multiple code elements on the same line

Run npm install to set up your project and then install debug@latest. <!-- ❌ -->

## File paths and commands

Look in the src/utils directory and run git clone https://github.com/user/repo.git <!-- ❌ -->

## Multiple elements spread across the document

Edit config.yaml in the settings/ folder <!-- ❌ -->

Run ./build.sh to compile the project <!-- ❌ -->

Add import os at the top of your script <!-- ❌ -->

## Nested code elements (should only fix outer ones)

The command \`echo PATH=/usr/bin\` sets your path variable to /usr/bin <!-- ❌ -->

## Elements after punctuation

Open readme.md, config.ini, and .env for configuration. <!-- ❌ -->
",
}
`;

exports[`Autofix snapshot tests backtick-code-elements rule backtick-code-elements-minimal autofix snapshot 1`] = `
{
  "fixed": "# Minimal auto-fix fixture

Install \`npm install\` <!-- ❌ -->

Go to \`src/utils\` <!-- ❌ -->

Edit \`config.yaml\` <!-- ❌ -->
",
  "hasChanges": true,
  "original": "# Minimal auto-fix fixture

Install npm install <!-- ❌ -->

Go to src/utils <!-- ❌ -->

Edit config.yaml <!-- ❌ -->
",
}
`;

exports[`Autofix snapshot tests sentence-case-heading rule sentence-case-heading autofix snapshot 1`] = `
{
  "fixed": "# Auto-fix fixture for sentence-case-heading

# This heading should be fixed <!-- ❌ -->

# Another heading should also fail <!-- ❌ -->

# Single word heading <!-- ❌ -->

# Heading with \`Code\` example <!-- ❌ -->
",
  "hasChanges": true,
  "original": "# Auto-fix fixture for sentence-case-heading

# this heading should be fixed <!-- ❌ -->

# Another Heading Should Also Fail <!-- ❌ -->

# SINGLE WORD HEADING <!-- ❌ -->

# heading with \`Code\` Example <!-- ❌ -->
",
}
`;
